File: MN/Algoritmi/Ecuatii-Neliniare/Secant_Method.m
function [x] = Secant_Method(xn_2, xn_1, tol, p, max_iter)
    % Solves f(x) = 0 by doing max_iter steps of the secant method
    % the secant method requires two initial values, xn_1 and xn_2
    % which should ideally be chosen close to the root
    % y = polyval(p, x) returns the value of a polynomial evaluated at x
    for i = 1 : max_iter
        % function values
        f_xn_2 = polyval(p, xn_2);
        f_xn_1 = polyval(p, xn_1);

        % the root of function f is approximated using the formula 
        xn = xn_2 - f_xn_2 * (xn_2 - xn_1) / (f_xn_2 - f_xn_1);

        % calculate f(xn)
        f_xn = polyval(p, xn);

        % xn is the solution
        if f_xn == 0
            x = xn;
            return;
        endif

        % calculate eps
        eps = abs((xn - xn_2) / xn);

        % is achieved a sufficient degree of accuracy eps
        if eps < tol
            x = xn;
            return;
        endif

        % the new values for the next step
        xn_1 = xn_2;
        xn_2 = xn;
    endfor
endfunction

File: python/nonlinear-equations/SecantMethod.py
def secant_method(f, x0, x1, tol, max_iter):
    # solves f(x) = 0 by doing max_iter steps of the secant method;
    # the secant method requires two initial values, x0 and x1
    # which should ideally be chosen close to the root;
    # y = feval(f, x)  evaluates a function using its name or its handle
    # and using the input arguments;

    eps = tol
    for i in range(1, max_iter):
        # function values: f(x0) and f(x1)
        f0 = f(x0)
        f1 = f(x1)

        # the root of function f is approximated using the formula
        xi = x1 - f1 * (x1 - x0) / (f1 - f0)

        # calculate f(xi)
        fi = f(xi)

        # xi is the solution
        if abs(fi) < eps:
            x = xi
            return x, i

        # calculate eps
        epsilon = abs((xi - x1) / xi)

        # stop if the secant method reached its convergence limit
        if epsilon < tol:
            x = xi
            return x, i

        # update the last two computed values
        x0 = x1
        x1 = xi
